lightfm/lightfm.py

```diff
@@ -601,20 +606,26 @@ def _run_epoch(self, item_features, user_features, interactions,
 
         # Call the estimation routines.
         if loss == 'warp':
-            fit_warp(CSRMatrix(item_features),
-                     CSRMatrix(user_features),
-                     positives_lookup,
-                     interactions.row,
-                     interactions.col,
-                     interactions.data,
-                     sample_weight,
-                     shuffle_indices,
-                     lightfm_data,
-                     self.learning_rate,
-                     self.item_alpha,
-                     self.user_alpha,
-                     num_threads,
-                     self.random_state)
+
+            # Cycle over scopes
+            for i in range(scope.shape[0]):
+                scope_item_ids = scope[0].indices
```

Comment:
I'm questioning whether we have to just have one scope of all 'active' recipes over the training data horizon instead of a scope per menu. The reason why is that for a particular menu scope the negative item ids might be correct, but the positive item ids are still sampled from all positive interactions over the horizon, so it might still result in invalid inversions.